Make JIT<->EE printing methods consistent and support nested classes when printing class names

jakobbotsch opened the pull request:
* Switch `appendClassName`/`getClassName`, `getFieldName` and `getMethodName` into respectively `printClassName`, `printFieldName` and `printMethodName` that all are consistent with `printObjectDescription` in buffer handling, and all use UTF8
* Change `printClassName` to support nested classes, fixing the handling of these in the JIT (in particular for method sets)
* Factor handling in SPMI/ILC/crossgen2 for these methods now that they are consistent
* Add a JitTypeNameFormatter.cs for crossgen2/ILC that does exactly what `printClassName` needs to be consistent with the VM
* Remove concept of "native" `CORINFO_METHOD_HANDLE` (`eeMarkNativeTarget` and co.). This was unused before.
* Remove `CEEInfo::getHelperName`. We were keeping a list of this in the JIT anyway under `FEATURE_SIMD`, which is defined practically everywhere.
* Remove `includeNamespaces` in JIT printing that was always passed as `true`
* Change the `hackishX` names on missing SPMI data into `<unknown X>` instead, e.g. `<unknown method>`

The change does complicate the JIT side slightly since the JIT owns all the buffers now. I have added a `buffer` argument to the various JIT helper methods (e.g. `eeGetFieldName`) that ensures we don't unnecessarily allocate memory in common short cases when printing these, so there are a few additional parameters passed, but in general I don't think it looks too bad. In a few recursive functions I have introduced a small lambda to avoid bloating the frame size by a lot (assuming it doesn't just get inlined).
I personally prefer the consistency and flexibility over the slight increase in complexity.

jakobbotsch committed:
Clean up JIT<->EE "get class name" methods

* Switch JIT to use appendClassName instead of getClassNameFromMetadata.
  This way it supports enclosing classes too when printing class names.
* Switch appendClassName to UTF8
* Switch SIMD type recognition to use getTypeInstantiationArgument and
  getClassNameFromMetadata, to use of appendClassName (and depending on
  how it formats instantiations)
* Add TypeString::FormatNoInst to avoid formatting instantiations, and
  use this for appendClassName

jakobbotsch committed:
Further cleanups/fixes

* Remove getClassName from JIT-EE interface
* Remove TypeString.cs in favor of JitTypeNameFormatter that does
  exactly what we need
* Bump JIT-EE GUID

jakobbotsch committed:
Run jit-format

SingleAccretion commented on src/coreclr/jit/simd.cpp:
```diff
@@ -290,143 +292,103 @@ CorInfoType Compiler::getBaseJitTypeAndSizeOfSIMDType(CORINFO_CLASS_HANDLE typeH
-                {
-                    size = getSIMDVectorRegisterByteLength();
+                CORINFO_CLASS_HANDLE typeArgHnd = info.compCompHnd->getTypeInstantiationArgument(typeHnd, 0);
+                simdBaseJitType                 = info.compCompHnd->asCorInfoType(typeArgHnd);
```
> `asCorInfoType`

Will the case of `Vector<MyEnum>` be rejected properly?

jakobbotsch replied:
I probably need `getTypeForPrimitiveNumericClass` instead.

jakobbotsch replied:
In fact, I can see that the code below for hw intrinsics is doing exactly this but with `getTypeForPrimitiveNumericClass`, so looks like that's definitely the way to go.

jakobbotsch commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/ObjectWriter/OutputInfoBuilder.cs:
```diff
@@ -182,7 +182,7 @@ public bool FindSymbol(OutputItem item, out int index)
         {
             DebugNameFormatter nameFormatter = new DebugNameFormatter();
-            TypeNameFormatter typeNameFormatter = new TypeString();
+            TypeNameFormatter typeNameFormatter = new JitTypeNameFormatter();
```
Hmm, didn't notice this use. Looks like I'll need to revert that change.

jakobbotsch committed:
Revert TypeString removal

jakobbotsch committed:
Use getTypeForPrimitiveNumericClass instead

jakobbotsch committed:
Nit

jakobbotsch committed:
Fix unsupported test

Apparently caller expects the size to be right even for unsupported
SIMD types.

jkotas commented on src/coreclr/tools/aot/ILCompiler.RyuJit/ILCompiler.RyuJit.csproj:
```diff
@@ -68,6 +68,9 @@
     <Compile Include="..\..\Common\JitInterface\TypeString.cs">
```
TypeString.cs is not used by the JIT/EE interface after this change, so it should not be in the JitInterface directory.

The one place that uses it is `OutputInfoBuilder` so it should be next to it.

jkotas commented on src/coreclr/jit/ee_il_dll.cpp:
```diff
@@ -1581,38 +1580,32 @@ const char16_t* Compiler::eeGetShortClassName(CORINFO_CLASS_HANDLE clsHnd)
-            constexpr bool fAssembly  = false;
-
+            int len = 0;
             // Warning: crossgen2 doesn't fully implement the `appendClassName` API.
```
Is this warning comment still relevant?

jakobbotsch committed:
Address feedback

jakobbotsch committed:
Remove leftover file

jkotas commented:
Test failing with:
```
Assert failure(PID 4100 [0x00001004], Thread: 14536 [0x38c8]): Assertion failed 'sizeBytes == 0' in 'JIT.HardwareIntrinsics.Arm.LoadPairScalarVector64_Int32:RunBasicScenario():this' during 'Importation' (IL size 78; hash 0x426d4f3e; FullOpts)
```

jakobbotsch commented:
Yep, I'm looking into it. The other failure looks like #48798.

jakobbotsch committed:
Fix test failure

jakobbotsch commented:
It's very odd that my PR has failed twice with the #48798 symptom now, and that issue has no other failures. I'll try to see if I can repro this tomorrow.

jakobbotsch committed:
Optimize

* Avoid multiple calls into EE in common case of short strings
* Remove StringPrinter::AllocToPrint complexity in face of above
* Optimize appendClassName on EE side with inlined implementation of
  what we need from TypeString that avoids multiple roundtrips in and
  out of UTF8. Also copy resulting string directly into final buffer.
* Revert TypeString::FormatNoInst

jakobbotsch committed:
Remove StringPrinter::Grow

jakobbotsch committed:
Fix contract, fix compiler error on x64

jakobbotsch committed:
Fix escaping

jakobbotsch committed:
Fix GCC build

jakobbotsch committed:
Clarify comment

jakobbotsch committed:
Avoid unnecessary eeGetMethodName call

We only use this name for helpers and it is actually very expensive to
ask the EE for it due to generic instantiations.

jakobbotsch committed:
Nit

jakobbotsch commented:
It seems #48798 was being caused in my PR by a very inefficient implementation of `appendClassName` that made JITting significantly slower in checked builds (around 50% more time spent when PMI'ing SPC). I am not completely sure why that would this particular test would be impacted more by the change, but it seems like it might be close to its timeout already since #48798 has intermittent failures.

The printing the JIT does here is very hot in checked builds since we are printing full signatures and also print class names for all struct layouts we see (in `ClassLayout`). Additionally, JIT was calling `appendClassName` twice even in the common case of short class names where it should be unnecessary.

I have fixed the unnecessary EE calls and pushed a new implementation of `appendClassName` that builds the class name directly into the buffer provided by the JIT without doing any redundant conversions or copies. The next step is to double check that crossgen2 with checked JIT is not significantly slower with the change.

jakobbotsch committed:
Remove StringPrinter::Printf in favor of Append

__vsnprintf_s seems to allocate heap memory on Windows and ends up
taking a lock. This can result in a ton of contention with checked JIT.
The worst part is that we do not even need the format strings, so just
get rid of it and use a simple memcpy to append.

With this the PR makes crossgen2 with checked JIT around 2 seconds
faster on my 5950X (32 threads, time goes from 7.5 seconds -> 5.5
seconds).

jakobbotsch commented:
The change actually made crossgen2 with checked JIT significantly faster. It took me a while to figure it out, but turns out there is a ton of contention in `__vsnprintf_s`, and the PR was removing one such call (printing the namespace). I pushed a commit to remove the rest of these calls as they are not even necessary, we can use simple `Append` instead. The result is that crossgen2 with checked JIT is around 2 seconds faster for SPC on my 5950X (32 threads, 7.5 seconds -> 5.5 seconds).

jakobbotsch committed:
Run jit-format

jakobbotsch commented:
I'm going to split this up into a few PRs instead.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into print-nested-classes

jakobbotsch committed:
Fix build

jakobbotsch committed:
appendClassName -> printClassName

jakobbotsch committed:
Nit

jkotas commented on src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -1890,37 +1882,26 @@ private CorInfoType asCorInfoType(CORINFO_CLASS_STRUCT_* cls)
             var type = HandleToObject(cls);
-            string name = TypeString.Instance.FormatName(type);
+            string name = JitTypeNameFormatter.Instance.FormatName(type);
+            byte[] utf8 = Encoding.UTF8.GetBytes(name);
```
This can use `Utf8.FromUtf16` + `Encoding.UTF8.GetByteCount` to avoid the allocation (it is the pattern used by `printObjectDescription`).

jakobbotsch replied:
Will factor these.

Does it make sense to make `getFieldName` and `getMethodName` consistent with the new style? It will require some more EE calls to replace the `scopeName` that they are returning additionally today. I suppose it shouldn't matter much since we should not be using them in release anyway.

jkotas replied:
I agree that the `scopeName` argument does not look pretty. Are you thinking about deleting the `scopeName` argument and switching the callers to use the `printClassName` instead?

jakobbotsch replied:
Yes, the callers would switch to `getFieldClass`/`getMethodClass` + `printClassName`  instead.

jakobbotsch committed:
Factor PrintFromUtf16

jakobbotsch committed:
Remove unnecessary function

jakobbotsch committed:
Fix missing write

jakobbotsch committed:
Work on switching getMethodName and getFieldName to print

jakobbotsch committed:
More work

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into print-nested-classes

jakobbotsch committed:
Fix build after merge

jakobbotsch committed:
Remove escaping type names complication

jakobbotsch committed:
Fix bad merge

jakobbotsch committed:
Fix bad merge

jakobbotsch committed:
Update for consistency, builds again

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
SPMI related fixes

jakobbotsch committed:
Unnecessary diff

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Avoid unnecessary diff 2

jakobbotsch committed:
Minor clean up for isEnum

jakobbotsch committed:
Avoid another diff

jakobbotsch committed:
Remove a couple more hackishX in favor of <unknown X>

jakobbotsch committed:
More consistent buffer size

jakobbotsch committed:
GC_NOTRIGGER

jakobbotsch committed:
Remove getHelperName

jakobbotsch committed:
Few fixes

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into print-nested-classes

jakobbotsch committed:
Handle a few more cases in JitTypeNameFormatter

jakobbotsch commented on src/coreclr/tools/superpmi/mcs/verbildump.cpp:
```diff
@@ -931,13 +937,15 @@ void DumpIL(MethodContext* mc)
     printf("// ProcessName - '%s'\n", mc->cr->repProcessName());
     printf(".assembly extern mscorlib{}\n");
-    printf(".assembly %s{}\n", moduleName);
+    printf(".assembly dumped_asm\n");
```
This "scope name" was not actually the module name.

jakobbotsch commented on src/coreclr/jit/gentree.cpp:
```diff
@@ -11942,7 +11937,11 @@ void Compiler::gtDispTree(GenTree*     tree,
         if (tree->OperIs(GT_FIELD, GT_FIELD_ADDR))
         {
-            printf(" %s", eeGetFieldName(tree->AsField()->gtFldHnd), 0);
+            auto disp = [&]() {
+                char buffer[256];
+                printf(" %s", eeGetFieldName(tree->AsField()->gtFldHnd, true, buffer, sizeof(buffer)));
+            };
+            disp();
```
Just avoiding bloating the stack frame here, since `gtDispTree` is recursive (well, it is left up to the C++ compiler's inlining).

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -3622,8 +3622,6 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
                                             info.compCompHnd->getFieldInClass(nullableCls, 0);
 
                                         assert(info.compCompHnd->getFieldOffset(hasValueFldHnd) == 0);
-                                        assert(!strcmp(info.compCompHnd->getFieldName(hasValueFldHnd, nullptr),
-                                                       "hasValue"));
```
We could print this (or better yet, introduce `getFieldNameFromMetadata`), but I don't think it's worth the effort to have the assert.

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -3417,118 +3417,107 @@ void CEEInfo::ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind entr
-        (*pnBufLen) -= nLenToCopy - 1;
+    if (pRequiredBufferSize != nullptr)
+    {
+        *pRequiredBufferSize = requiredBufferSize;
```
Do we need `+ 1` here for the null terminator?

jakobbotsch replied:
Yep, good catch.

```suggestion
        *pRequiredBufferSize = requiredBufferSize + 1;
```

jkotas reviewed (approved):
VM/AOT parts LGTM

jakobbotsch committed:
Update src/coreclr/vm/jitinterface.cpp

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/asmdumper.cpp:
```diff
@@ -19,8 +19,11 @@ void ASMDumper::DumpToFile(HANDLE hFile, MethodContext* mc, CompileResult* cr)
     buff_offset += sprintf_s(&buff[buff_offset], bufflen - buff_offset,
                              ";;Generated from SuperPMI on original input '%s'", cr->repProcessName());
+
+    char methodName[256];
```
I'm a little worried about all these constant `256` around. Can we use something like `printInFull` in these places that automatically get the full string? We need something like a C++ holder helper class to automatically delete memory, or maybe a way to associate the memory with a MethodContext.

jakobbotsch replied:
I introduced `getMethodName` and `getClassName` in spmiutil.cpp that use `std::string` and print the full strings.

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+            res.requiredBufferSize = requiredBufferSize;
+        }
+
+        if (bytesWritten > res.bufferSize)
```
```suggestion
        if (bytesWritten >= res.bufferSize)
```
bytesWritten doesn't include null character, but bufferSize does

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+        {
+            // Always stored without null terminator.
+            res.buffer = map->AddBuffer((unsigned char*)buffer, static_cast<unsigned>(bytesWritten));
+            res.bufferSize = (UINT)bytesWritten;
```
```suggestion
            res.bufferSize = (UINT)(bytesWritten + 1);
```
?

BruceForstall replied:
There's something confusing here where you aren't actually storing the argument `bufferSize`, which makes sense, but what you are storing as `bufferSize` doesn't have quite the same semantics. Maybe the stored `bufferSize` should get a different name, such as `bufferBytesWritten`?

jakobbotsch replied:
I can understand the confusion. The `Agnostic_PrintResult::bufferSize` is actually just the size of the buffer that `Agnostic_PrintResult::buffer` is pointing at. So it is not really related to the buffer size passed to these functions.
The string is always stored without null terminator.

With the unified string handling, I've spent the extra effort to implement this in a way where we don't need to split the key on the buffer size, and simply always store the "maximal" information we queried during recording. The benefit is that we are able to replay for any `bufferSize` below that (or above it, if we have all data) without missing.

I think to make this less confusing I will rename to `Agnostic_PrintResult::stringBuffer` and `Agnostic_PrintResult::stringBufferSize` to avoid the conflict with the `bufferSize` parameter in these function. With that in mind, does the string handling make more sense in regards to the null terminators?

BruceForstall replied:
> I think to make this less confusing I will rename to `Agnostic_PrintResult::stringBuffer` and `Agnostic_PrintResult::stringBufferSize` to avoid the conflict with the `bufferSize` parameter in these function. With that in mind, does the string handling make more sense in regards to the null terminators?

Yes, as long as there are good comments that specify if the `stringBufferSize` includes the null terminator or not.

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
-    // Don't save the string buffer if the incoming buffer length is zero, even if the string buffer is non-null.
-    // In that case, the EE/crossgen2 don't zero-terminate the buffer (since they assume it is zero sized).
-    if ((result != nullptr) && (nBufLenIn > 0))
+        dmpPrint(name, map, handle, res);
```
```suggestion
        DEBUG_REC(dmpPrint(name, map, handle, res));
```

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
+    res.requiredBufferSize = requiredBufferSize;
+
+    map->Add(handle, res);
+    dmpPrint(name, map, handle, res);
```
```suggestion
    DEBUG_REC(dmpPrint(name, map, handle, res));
```

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shared/methodcontext.cpp:
```diff
@@ -6683,117 +6443,174 @@ CORINFO_CLASS_HANDLE MethodContext::repGetTypeInstantiationArgument(CORINFO_CLAS
 
-    AppendClassName->Add(key, value);
-    DEBUG_REC(dmpAppendClassName(key, value));
+    res.bufferSize = (UINT)bytesWritten;
```
```suggestion
    res.bufferSize = (UINT)(bytesWritten + 1);
```
?

BruceForstall commented on src/coreclr/jit/eeinterface.cpp:
```diff
@@ -226,21 +221,39 @@ void Compiler::eePrintMethod(StringPrinter*        printer,
 
-    const char* methName = info.compCompHnd->getMethodName(methHnd, nullptr);
-    printer->Append(methName);
+    size_t requiredBufferSize;
+    char   buffer[256];
+    info.compCompHnd->printMethodName(methHnd, buffer, sizeof(buffer), &requiredBufferSize);
+    if (sizeof(buffer) <= requiredBufferSize)
+    {
+        printer->Append(buffer);
+    }
+    else
+    {
+        char* pBuffer = new (this, CMK_DebugOnly) char[requiredBufferSize];
+        info.compCompHnd->printMethodName(methHnd, pBuffer, requiredBufferSize);
+        printer->Append(pBuffer);
+    }
```
This is a common pattern now. Might be nice to abstract it.

jakobbotsch replied:
Done, I added an `eeAppendPrint` function to do this. Also realized the `sizeof(buffer) <= requiredBufferSize` check is flipped.

jakobbotsch committed:
Address feedback, remove outdated comment I forgot to remove in previous PR

jakobbotsch committed:
Address some more feedback

jakobbotsch commented on src/coreclr/inc/corinfo.h:
```diff
@@ -2282,7 +2282,7 @@ class ICorStaticInfo
     //
     // Return Value:
-    //    Bytes written to the given buffer, the range is [0..bufferSize)
+    //    Bytes written to the given buffer excluding the null terminator. The range is [0..bufferSize).
```
Can you please take a look at the extra docs I've added in the latest commits?

jakobbotsch committed:
Small fix

BruceForstall reviewed (approved):
LGTM
